xray/default_emitter.go

```diff
@@ -57,6 +69,13 @@ func (de *DefaultEmitter) Emit(seg *Segment) {
 			return b.String()
 		})
 		de.Lock()
+
+		if de.conn == nil {
```

Comment:
Should we do this check before `packSegments` call. In high intensive subsegment streaming, the current check will be performed for each batch (though only first check will dial up), or if dial up throws error, we dont parse the tree : small optimization. 
Also in case of error 

```
if err := de.refresh(de.addr); err != nil {
				return
			}
```
I think we should first release the lock before returning, so that subsequent operations acquire the lock.

Reply:
I fixed the missing `Unlock`.

As for the order. I think it is better to put it inside the loop so we don't have to lock and unlock the mutex an additional time prior to the loop. That will improve performance somewhat for the "normal" case, which I think makes more sense than optimizing performance for an error case.

Reply:
Sure, fair point.